pallets/dedir/src/lib.rs

```diff
@@ -0,0 +1,456 @@
+			who: T::AccountId,
+			registry_id: RegistryIdOf<T>,
+			registry_entry_id: RegistryEntryIdOf<T>,
+		},
+		/// State change has been made for existing registry entry.
+		/// \[who, registry_identifier, registry_entry_identifier, new_state\]
+		RegistryEntryStateChanged {
+			who: T::AccountId,
+			registry_id: RegistryIdOf<T>,
+			registry_entry_id: RegistryEntryIdOf<T>,
+			new_state: RegistryStateOf<T>,
+		},
+	}
+
+	#[pallet::call]
+	/// DeDir pallet declaration.
+	impl<T: Config> Pallet<T> {
+		/// A new Registry has been created.
+		///
+		/// This function allows a user to submit a new create registry request.
+		/// The Registry is created along with various metadata, including the
+		/// attributes, additional-states to be supported.
+		///
+		/// # Arguments
+		/// * `origin` - The origin of the call, which should be a signed user in most cases.
+		/// * `registry_id` - Registry Identifier to be associated with the Registry.
+		/// * `attrs` - (Optional) Bounded Vector of RegistryKeys and associated RegistryTypes which
+		///   will be supported.
+		/// * `additional_states` - (Optional) Bounded Vector of Additional States to be supported
+		///   by the Registry.
+		///
+		/// # Errors
+		/// Returns `Error::<T>::RegistryIdAlreadyExists` if the registry identifier
+		/// exists already.
+		/// Returns `Error::<T>::MaxRegistryStorageOverflow` if the Registry Attributes
+		/// storage upper bound has been breached.
+		/// Returns `Error::<T>::AddingStateFailed` if the state addition fails to be added.
+		///
+		/// # Events
+		/// Emits `CreatedRegistry` when a new Registry is created successfully.
+		///
+		/// # Example
+		/// ```
+		/// create_registry(origin, registry_id, attrs, additional_states)?;
+		/// ```
+		#[pallet::call_index(0)]
+		#[pallet::weight({0})]
+		pub fn create_registry(
+			origin: OriginFor<T>,
+			registry_id: RegistryIdOf<T>,
+			attrs: Vec<(RegistryKeyIdOf<T>, RegistrySupportedTypeOf)>,
+			// TODO: Use Runtime Constant for upper bound
+			additional_states: Option<BoundedVec<BoundedVec<u8, ConstU32<32>>, ConstU32<10>>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			/* Ensure that the registry_id does not already exist */
+			ensure!(
+				!Registries::<T>::contains_key(&registry_id),
+				Error::<T>::RegistryIdAlreadyExists
+			);
+
+			/* Set registry with default states */
+			let mut registry = Registry::new();
+
+			/* Ensure Registry attributes does not overflow */
+			for (registry_key, registry_key_type) in attrs {
+				let entry = Entry { registry_key, registry_key_type };
+				if registry.entries.try_push(entry).is_err() {
+					return Err(Error::<T>::MaxRegistryStorageOverflow.into());
+				}
+			}
+
+			/* Ensure new-states does not overflow */
+			if let Some(states) = additional_states {
+				if registry.add_supported_states(states).is_err() {
+					return Err(Error::<T>::AddingStateFailed.into());
+				}
+			}
+
+			Registries::<T>::insert(&registry_id, registry);
```

Comment:
registry ownership?

Reply:
Delegation/Ownership is not implemented in this version. Will do with next commit.